### DISCUSSION TITLE
Integrate Octopus Agile API to target cheapest overnight rates (Get paid to charge)

### DISCUSSION THREAD

**Discussion by kme77** [nathanmarlor/foxess_em#76] [Ideas]
I realise I know very little, but I thought I would put this out there as I can't see it being mentioned here.

Would it be possible to integrate the Octopus Agile API in order to only charge during the cheapest hours during the upcoming 24 hours? I assume this would work much better once Modbus input controls are mapped on the inverter avoiding use of the Foxcloud. I feel this would be useful for everyone on the Agile tariff, even those with just battery storage and no solar?

I feel this could be used to further granularize the charging system in that once the total required charge for the upcoming day is known, the charging period(s) could be spread over the next day during the cheapest 30 min slots, starting with the very cheapest then moving to the next cheapest and so on until fully charged.

![agile api](https://user-images.githubusercontent.com/38857063/201318962-9f1bb3b9-0189-4b21-9641-ae6a2e4d4626.JPG)

**Comment by kme77** [comment]
https://developer.octopus.energy/docs/api/#list-tariff-charges

**Comment by kme77** [comment]
As an example, checking the Agile prices for today https://agileprices.co.uk// it would have been possible to charge sporadically between the hours of 1.30-2am, 2.30-3am and 4-5.30am and as the prices went NEGATIVE during these times, instead of paying to charge the battery we could have been PAID @ an average of 4.72p/kW for 2.5hrs.
I realise that this is an extreme example, but I think it demonstrates the concept quite nicely! :)

![paid to charge](https://user-images.githubusercontent.com/38857063/201327686-fc2b8f7c-1278-4521-91ab-27bf4615bcdb.JPG)

**Comment by nathanmarlor** [comment]
This looks great, thanks @kme77!

I'll digest with a coffee on Monday morning and see what's possible :+1:

**Reply by kme77** [reply]
No problem, just happy to contribute in any way I can no matter how small. Very nice work btw! 👍

I went digging and found this component from markgdev which looks like it could do what I was suggesting. Apparently it's no longer being maintained (though the dev was active this week) so maybe there is the option to take over/integrate it with yours?

At the very least it could act as a starting point. ⚡

https://github.com/markgdev/home-assistant_OctopusAgile

I hope I'm not being presumptuous? 😬

Thanks again for the awesome contribution!

**Reply by nathanmarlor** [reply]
Think I've got the concept straight now...!

A quick thought though, on agileprices.co.uk most prices seem to be ~20p/kW - surely it's cheaper to be on Octopus Go @ 7.5p/kW? Does this improve in summer?

**Reply by kme77** [reply]
Unfortunately I believe Octopus Go/Intelligent are only available to EV owners so if like myself you're stuck with an ICE car (for now) Agile is the only option (other than Eco7)

Side note, Agile isn't currently available to sign up (hopefully they will bring it back soon) so I've signed up to eco7 in the meantime with a 17p/kWh 0:00-7.00am rate which might actually end up being a better option if Agile ends up averaging out higher.

**Reply by smoketabs** [reply]
Hi. Just wanted to ask about any progress/thoughts on the EM integration making use of Agile for import?

Currently using with Flux and recently upgraded cube stack to 24kWh (mostly for dark months).

Thanks.

**Reply by icornish72** [reply]
As a recent convert to solar PV + battery with a Foxess battery and inverter, I would be keen to look into this greater.
I currently have a simple force charge automation that if the unit price is less than 15p, and it's after sunset/before sunrise, and the battery is less than 50% full, then charge for the next 30 minutes.. 
Sadly in my 3 weeks on agile/PV/battery, I've not yet seen that behaviour actually happen, but I wait with baited breathe!

Would love to see some sort of integration with the octopus agile API on https://github.com/markgdev/home-assistant_OctopusAgile which seems to have some functions to do something when the prices are low - am just working on an automation to leverage this.
